### PR TITLE
Define types for struct fields and class methods

### PR DESCRIPTION
**Pull request by jbouwman**
Refactor: make modeling of structs and classes more consistent by combining parallel slots into structures.

- replace 3 parallel hashtables in struct-entry with list of struct-field (name, type, index, docstring)
- replace parallel lists of (name . type) and docstring in ty-class with list of ty-class-method (name, type, docstring)

### CODE REVIEW

File: src/typechecker/environment.lisp [hunk 1]
```diff
@@ -496,17 +505,26 @@
...
-  ;; Mapping of "field name" -> "field type"
-  ;; Type variables are the same as in `type-entry-type'
-  (field-tys        (util:required 'field-tys)        :type environment-map  :read-only t)
+(defun struct-field-list-p (x)
+  (and (alexandria:proper-list-p x)
+       (every #'struct-field-p x)))
 
-  ;; Mapping of "field name" -> "field index"
-  (field-idx        (util:required 'field-idx)        :type environment-map  :read-only t))
+(deftype struct-field-list ()
```
Comment (Izaakwltn) [inline src/typechecker/environment.lisp:521]: :thumbsup:

File: src/codegen/translate-instance.lisp [hunk 2]
```diff
@@ -113,4 +116,4 @@
           :do (typecheck-node node env))
 
     bindings))
- 
+
```
Comment (Izaakwltn) [inline src/codegen/translate-instance.lisp:119]: true stupid nit

File: src/doc/generate-documentation.lisp [hunk 3]
```diff
@@ -25,9 +25,7 @@
             (:include documentation-entry))
   (type             (util:required 'type)             :type tc:ty                     :read-only t)
   (tyvars           (util:required 'tyvars)           :type tc:tyvar-list             :read-only t)
-  (fields           (util:required 'fields)           :type util:string-list          :read-only t)
-  (field-docstrings (util:required 'field-docstrings) :type hash-table                :read-only t)
-  (field-tys        (util:required 'field-tys)        :type hash-table                :read-only t)
+  (fields           (util:required 'fields)           :type tc:struct-field-list      :read-only t)
```
Comment (Izaakwltn) [inline src/doc/generate-documentation.lisp:28]: Have you tried building documentation locally to make sure the docstrings are still transmitted?

**Review by Izaakwltn — APPROVED** [review]
LGTM- more readable and explicit.

File: src/doc/generate-documentation.lisp (hunk 3 above)
Comment (jbouwman) [reply src/doc/generate-documentation.lisp:28]: Hm, nope, good point. It built and tested clean, so I pushed and immediately went for a cheerful walk around the block.

I've now confirmed that docstrings are still propagated.

**Review by eliaslfox — APPROVED** [review]

**Pull request merged**